File: python/Leetcode-Roadmap/01_Array_And_Hashing/valid_sudoku.py
import collections

class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        # Initialize dictionaries to keep track of numbers in each row, column, and square
        cols = collections.defaultdict(set)
        rows = collections.defaultdict(set)
        squares = collections.defaultdict(set)  # key = (r // 3, c // 3)

        # Iterate through each cell in the Sudoku board
        for r in range(9):
            for c in range(9):
                # Skip empty cells
                if board[r][c] == ".":
                    continue

                # Check if the current number is already present in the same row, column, or square
                if (
                    board[r][c] in rows[r]
                    or board[r][c] in cols[c]
                    or board[r][c] in squares[(r // 3, c // 3)]
                ):
                    return False

                # Add the current number to the sets for the corresponding row, column, and square
                cols[c].add(board[r][c])
                rows[r].add(board[r][c])
                squares[(r // 3, c // 3)].add(board[r][c])

        # If no conflicts were found, the Sudoku board is valid
        return True
    

# implementation
print("Leetcode 36: Valid Sudoku\n")
answer = Solution()

# Example 1
example_1 = [["5","3",".",".","7",".",".",".","."]
,["6",".",".","1","9","5",".",".","."]
,[".","9","8",".",".",".",".","6","."]
,["8",".",".",".","6",".",".",".","3"]
,["4",".",".","8",".","3",".",".","1"]
,["7",".",".",".","2",".",".",".","6"]
,[".","6",".",".",".",".","2","8","."]
,[".",".",".","4","1","9",".",".","5"]
,[".",".",".",".","8",".",".","7","9"]]
print(f"Example 1: board = {example_1}")
print("Output:", answer.isValidSudoku(example_1))

# Example 2
example_2 = [["8","3",".",".","7",".",".",".","."]
,["6",".",".","1","9","5",".",".","."]
,[".","9","8",".",".",".",".","6","."]
,["8",".",".",".","6",".",".",".","3"]
,["4",".",".","8",".","3",".",".","1"]
,["7",".",".",".","2",".",".",".","6"]
,[".","6",".",".",".",".","2","8","."]
,[".",".",".","4","1","9",".",".","5"]
,[".",".",".",".","8",".",".","7","9"]]
print(f"Example 2: board = {example_2}")
print("Output:", answer.isValidSudoku(example_2))

File: python/Leetcode-Roadmap/01_Array_And_Hashing/valid_sudoku.js
// Function to check the validity of a Sudoku board
var isValidSudoku = (board) => {
    const [boards, cells] = [3, 9];
    const [boxes, rows, cols] = getBoards(boards, cells); /* Time O(ROWS * COLS) | Space O(CELLS) */

    return searchGrid(board, boxes, rows, cols); /* Time O(ROWS * COLS) | Space O(CELLS) */
}

// Helper function to initialize arrays for tracking numbers in each box, row, and column
var getBoards = (boards, cells) => new Array(boards).fill()
    .map(() => new Array(cells).fill(0));

// Helper function to search through the Sudoku grid and check for valid numbers
var searchGrid = (board, boxes, rows, cols) => {
    const [_rows, _cols] = [9, 9];

    for (let row = 0; row < _rows; row++) { /* Time O(ROWS)*/
        for (let col = 0; col < _cols; col++) { /* Time O(COLS)*/
            const char = board[row][col];
            const position = 1 << (char - 1);
            const index = (Math.floor(row / 3) * 3) + Math.floor(col / 3);

            const isEmpty = char === '.';
            if (isEmpty) continue;

            const hasMoved = (boxes[index] & position) || (cols[col] & position) || (rows[row] & position);
            if (hasMoved) return false;

            // Mark the position of the current number in the corresponding box, row, and column
            rows[row] |= position; /* Space O(CELLS)*/
            cols[col] |= position; /* Space O(CELLS)*/
            boxes[index] |= position; /* Space O(CELLS)*/
        }
    }

    // If the loop completes without finding conflicts, the Sudoku board is valid
    return true;
}

// implementation
console.log("Leetcode 36: Valid Sudoku\n");

// Example 1
example1 = [["5","3",".",".","7",".",".",".","."]
,["6",".",".","1","9","5",".",".","."]
,[".","9","8",".",".",".",".","6","."]
,["8",".",".",".","6",".",".",".","3"]
,["4",".",".","8",".","3",".",".","1"]
,["7",".",".",".","2",".",".",".","6"]
,[".","6",".",".",".",".","2","8","."]
,[".",".",".","4","1","9",".",".","5"]
,[".",".",".",".","8",".",".","7","9"]];

console.log(`Example 1: board = [${example1}]`);
console.log("Output:", isValidSudoku(example1));

// Example 2
example2 = [["8","3",".",".","7",".",".",".","."]
,["6",".",".","1","9","5",".",".","."]
,[".","9","8",".",".",".",".","6","."]
,["8",".",".",".","6",".",".",".","3"]
,["4",".",".","8",".","3",".",".","1"]
,["7",".",".",".","2",".",".",".","6"]
,[".","6",".",".",".",".","2","8","."]
,[".",".",".","4","1","9",".",".","5"]
,[".",".",".",".","8",".",".","7","9"]];

console.log(`Example 2: board = [${example2}]`);
console.log("Output:", isValidSudoku(example2));